Obtención de estadisticas descriptivas

Integrantes: 
Pablo Enrique Martinez Sanchez A01706352
Carlos Iñaki Román Martinez A01702712
Jesus Dassaef Lopez Barrios A01366815

Entendimiento de las columnas según el creador: 
Game ID;
Rated (T/F);
Start Time;
End Time;
Number of Turns;
Game Status;
Winner;
Time Increment;
White Player ID;
White Player Rating;
Black Player ID;
Black Player Rating;
All Moves in Standard Chess Notation;
Opening Eco (Standardised Code for any given opening, list here);
Opening Name;
Opening Ply (Number of moves in the opening phase)

link de consulta: https://www.kaggle.com/datasnaek/chess


En este programa analizaremos una base de datos de más de 20,000 partidas de ajedrez en la pagina de lichess.org
Se busca encontrar relaciones entre rating, color, resultado, movimientos de apertura, y movimientos hechos.
En la primera parte del código se encuentra información general de la base de datos y su entendimiento.

In [ ]:
import pandas as pd  #Importamos la libreria de pandas y la renombarmos para más facilidad
import numpy as np
from matplotlib import pyplot as plt #importamos la libreria para graficar 
data = pd.read_csv("gamesChess.csv") #Importamos el archivo con la información


In [ ]:
data.head() #Mostramos la información de nuestro archivo para corroborrar que está en orden


id  rated    created_at  last_move_at  turns victory_status winner  \
0  TZJHLljE  False  1.504210e+12  1.504210e+12     13      outoftime  white   
1  l1NXvwaE   True  1.504130e+12  1.504130e+12     16         resign  black   
2  mIICvQHh   True  1.504130e+12  1.504130e+12     61           mate  white   
3  kWKvrqYL   True  1.504110e+12  1.504110e+12     61           mate  white   
4  9tXo1AUZ   True  1.504030e+12  1.504030e+12     95           mate  white   

  increment_code       white_id  white_rating      black_id  black_rating  \
0           15+2       bourgris          1500          a-00          1191   
1           5+10           a-00          1322     skinnerua          1261   
2           5+10         ischia          1496          a-00          1500   
3           20+0  daniamurashov          1439  adivanov2009          1454   
4           30+3      nik221107          1523  adivanov2009          1469   

                                               moves opening_eco  \
0  d4 d5 c4 c6 cxd5 e6 dxe6 fxe6 Nf3 Bb4+ Nc3 Ba5...         D10   
1  d4 Nc6 e4 e5 f4 f6 dxe5 fxe5 fxe5 Nxe5 Qd4 Nc6...         B00   
2  e4 e5 d3 d6 Be3 c6 Be2 b5 Nd2 a5 a4 c5 axb5 Nc...         C20   
3  d4 d5 Nf3 Bf5 Nc3 Nf6 Bf4 Ng4 e3 Nc6 Be2 Qd7 O...         D02   
4  e4 e5 Nf3 d6 d4 Nc6 d5 Nb4 a3 Na6 Nc3 Be7 b4 N...         C41   

                             opening_name  opening_ply  
0        Slav Defense: Exchange Variation            5  
1  Nimzowitsch Defense: Kennedy Variation            4  
2   King's Pawn Game: Leonardis Variation            3  
3  Queen's Pawn Game: Zukertort Variation            3  
4                        Philidor Defense            5

In [ ]:
data.columns #mostramos las columnas de la tabla

Index(['id', 'rated', 'created_at', 'last_move_at', 'turns', 'victory_status',
       'winner', 'increment_code', 'white_id', 'white_rating', 'black_id',
       'black_rating', 'moves', 'opening_eco', 'opening_name', 'opening_ply'],
      dtype='object')

In [ ]:
data.dtypes #Mostramos el tipo de información que existe

id                 object
rated                bool
created_at        float64
last_move_at      float64
turns               int64
victory_status     object
winner             object
increment_code     object
white_id           object
white_rating        int64
black_id           object
black_rating        int64
moves              object
opening_eco        object
opening_name       object
opening_ply         int64
dtype: object

In [ ]:
data.info() #Mostramos el tipo de datos en cada columna

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20058 entries, 0 to 20057
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              20058 non-null  object 
 1   rated           20058 non-null  bool   
 2   created_at      20058 non-null  float64
 3   last_move_at    20058 non-null  float64
 4   turns           20058 non-null  int64  
 5   victory_status  20058 non-null  object 
 6   winner          20058 non-null  object 
 7   increment_code  20058 non-null  object 
 8   white_id        20058 non-null  object 
 9   white_rating    20058 non-null  int64  
 10  black_id        20058 non-null  object 
 11  black_rating    20058 non-null  int64  
 12  moves           20058 non-null  object 
 13  opening_eco     20058 non-null  object 
 14  opening_name    20058 non-null  object 
 15  opening_ply     20058 non-null  int64  
dtypes: bool(1), float64(2), int64(4), object(9)
memory usage: 2.3+ MB


In [ ]:
data .describe() #Mostramos la información estadistica

created_at  last_move_at         turns  white_rating  black_rating  \
count  2.005800e+04  2.005800e+04  20058.000000  20058.000000  20058.000000   
mean   1.483617e+12  1.483618e+12     60.465999   1596.631868   1588.831987   
std    2.850151e+10  2.850140e+10     33.570585    291.253376    291.036126   
min    1.376772e+12  1.376772e+12      1.000000    784.000000    789.000000   
25%    1.477548e+12  1.477548e+12     37.000000   1398.000000   1391.000000   
50%    1.496010e+12  1.496010e+12     55.000000   1567.000000   1562.000000   
75%    1.503170e+12  1.503170e+12     79.000000   1793.000000   1784.000000   
max    1.504493e+12  1.504494e+12    349.000000   2700.000000   2723.000000   

        opening_ply  
count  20058.000000  
mean       4.816981  
std        2.797152  
min        1.000000  
25%        3.000000  
50%        4.000000  
75%        6.000000  
max       28.000000

In [ ]:
pd.unique(data["rated"]) #Que valores existen en la columna rated

array([False,  True])

In [ ]:
data["rated"].describe() #Analísis estadistico de la columna, top nos muestra cual es la moda en la columna en este caso los jugadores registrados.

count     20058
unique        2
top        True
freq      16155
Name: rated, dtype: object

In [ ]:
data_size = data.size
print("Cantidad de datos :",data_size)

Cantidad de datos : 320928


In [ ]:
data.info # conocer las ramas y columnas

<bound method DataFrame.info of              id  rated    created_at  last_move_at  turns victory_status  \
0      TZJHLljE  False  1.504210e+12  1.504210e+12     13      outoftime   
1      l1NXvwaE   True  1.504130e+12  1.504130e+12     16         resign   
2      mIICvQHh   True  1.504130e+12  1.504130e+12     61           mate   
3      kWKvrqYL   True  1.504110e+12  1.504110e+12     61           mate   
4      9tXo1AUZ   True  1.504030e+12  1.504030e+12     95           mate   
...         ...    ...           ...           ...    ...            ...   
20053  EfqH7VVH   True  1.499791e+12  1.499791e+12     24         resign   
20054  WSJDhbPl   True  1.499698e+12  1.499699e+12     82           mate   
20055  yrAas0Kj   True  1.499698e+12  1.499698e+12     35           mate   
20056  b0v4tRyF   True  1.499696e+12  1.499697e+12    109         resign   
20057  N8G2JHGG   True  1.499643e+12  1.499644e+12     78           mate   

      winner increment_code       white_id  white_ratin

In [ ]:
pd.unique(data["turns"]) #Que valores existen en la columna turns

array([ 13,  16,  61,  95,   5,  33,   9,  66, 119,  39,  38,  60,  31,
        43,  52, 101,  25,  14,   3,  17,  36,  69,  54,  53,  64,  21,
        11,  19,  28,  22,  37,  34,  32,  71,  30,  41,  49,   7,  35,
         8,  44,  29,  40,  57,  78,  62,  87,  50, 131,  48,  47,  51,
        58, 114,  20,  85, 113,  75, 135,  81,  46,  70,  91, 137,  18,
       111,  90,  59,  26,  23, 118, 120,  24,   2,  97,  12,  72, 156,
        77,  65,  73, 125,  76,  63, 107, 100,  80, 105, 144, 158, 117,
       110, 178, 136, 124,  96,  10, 106,  15,  42, 159,  74, 112, 104,
       150,  88,  55, 103, 177,  82,  99,  67,   6, 115,  83,  89, 116,
        27, 102,  45,  98, 108,  92,  56, 123, 129,   4, 163, 130, 121,
        94,  79, 148, 195, 122, 128,  84,  86,  93, 161, 133,  68, 153,
       154, 160, 141, 132, 204, 140, 168, 127, 109, 174, 157, 142, 143,
       170,   1, 138, 171, 145, 151, 179, 149, 173, 208, 139, 134, 169,
       155, 164, 147, 126, 146, 190, 167, 255, 210, 176, 172, 18

In [ ]:
mediana_turns = data["turns"].median()
media_turns   = data["turns"].mean()
moda_turns    = data["turns"].mode().max()
std_turns     = data["turns"].std()
min_turns     = data["turns"].min()
max_turns     = data["turns"].max()

print("Mediana turns      :",mediana_turns)
print("Media turns        :",media_turns)
print("Moda turns         :",moda_turns)
print("Desv. Est. turns   :",std_turns)
print("Valor mínimo turns :",min_turns)
print("Valor máximo turns :",max_turns)

data["turns"].describe()

Mediana turns      : 55.0
Media turns        : 60.46599860404826
Moda turns         : 53
Desv. Est. turns   : 33.570584753537055
Valor mínimo turns : 1
Valor máximo turns : 349


count    20058.000000
mean        60.465999
std         33.570585
min          1.000000
25%         37.000000
50%         55.000000
75%         79.000000
max        349.000000
Name: turns, dtype: float64

¿Qué representa? -> La variable 'turns' representa el número de movimientos totales en una partida, en la cual, un movimiento representa tanto uno de blancas como uno de negras, es decir, cuando ambos jugadores han hecho sus movimientos (1 cada uno).

Conclusiones ->  Se denota que en las partidas de ajedrez de la pagina lichess.org lo más normal es que sean 60 movimientos, donde la moda es de 53, cabe denotar que la desviación estandar no se puede tomar de forma significativa ya que el maximo y el minimo están muy separados, para más información se deberia de revisar la tabla de distribución.

In [ ]:
pd.unique(data["winner"])

array(['white', 'black', 'draw'], dtype=object)

In [ ]:
# mediana_winner = data["winner"].median()
# media_winner   = data["winner"].mean()
moda_winner    = data["winner"].mode().max()
# std_winner     = data["winner"].std()
# min_winner     = data["winner"].min()
# max_winner     = data["winner"].max()

# print("Mediana winner      :",mediana_winner)
# print("Media winner        :",media_winner)
print("Moda winner         :",moda_winner)
# print("Desv. Est. winner   :",std_winner)
# print("Valor mínimo winner :",min_winner)
# print("Valor máximo winner :",max_winner)

data["winner"].describe()

Moda winner         : white


count     20058
unique        3
top       white
freq      10001
Name: winner, dtype: object

¿Qué representa? -> La variable winner nos da el atributo si la partida fue ganada por blancas, negras o un empate.



```
# Esto tiene formato de código
```

Conclusiones ->  Podemos ver que las personas que juegan de blancas y tienen el primer movimiento tienden a ganar la partida.

In [ ]:
pd.unique(data["victory_status"])

array(['outoftime', 'resign', 'mate', 'draw'], dtype=object)

In [ ]:
# mediana_victory_status = data["victory_status"].median()
# media_victory_status   = data["victory_status"].mean()
moda_victory_status    = data["victory_status"].mode().max()
# std_victory_status     = data["victory_status"].std()
# min_victory_status     = data["victory_status"].min()
# max_victory_status     = data["victory_status"].max()

# print("Mediana victory_status      :",mediana_victory_status)
# print("Media victory_status        :",media_victory_status)
print("Moda victory_status         :",moda_victory_status)
# print("Desv. Est. victory_status   :",std_victory_status)
# print("Valor mínimo victory_status :",min_victory_status)
# print("Valor máximo victory_status :",max_victory_status)

data["victory_status"].describe() 

Moda victory_status         : resign


count      20058
unique         4
top       resign
freq       11147
Name: victory_status, dtype: object

¿Qué representa? -> La variable victory_status nos dice en que terminaron las partidas, se puede terminar una partida por que se acabo el tiempo de un jugador, jaque mate, empate o que un participante renuncie.

Conclusiones ->  Se puede observar que las mayorias de las partidas son terminada por que uno de los juadores resigna, esto se puede ya que en ajedrez es común que si la partida está perdida la persona se rinda.

In [ ]:
pd.unique(data["white_rating"])

array([1500, 1322, 1496, ...,  860,  929,  880], dtype=int64)

In [ ]:
mediana_white_rating = data["white_rating"].median()
media_white_rating   = data["white_rating"].mean()
moda_white_rating    = data["white_rating"].mode().max()
std_white_rating     = data["white_rating"].std()
min_white_rating     = data["white_rating"].min()
max_white_rating     = data["white_rating"].max()

print("Mediana white_rating      :",mediana_white_rating)
print("Media white_rating        :",media_white_rating)
print("Moda white_rating         :",moda_white_rating)
print("Desv. Est. white_rating   :",std_white_rating)
print("Valor mínimo white_rating :",min_white_rating)
print("Valor máximo white_rating :",max_white_rating)

data["white_rating"].describe() 

Mediana white_rating      : 1567.0
Media white_rating        : 1596.6318675840064
Moda white_rating         : 1500
Desv. Est. white_rating   : 291.25337573701944
Valor mínimo white_rating : 784
Valor máximo white_rating : 2700


count    20058.000000
mean      1596.631868
std        291.253376
min        784.000000
25%       1398.000000
50%       1567.000000
75%       1793.000000
max       2700.000000
Name: white_rating, dtype: float64

¿Qué representa? -> Aqui podemos ver cuál es el puntaje de cada jugador que juegan con blancas, esto se ganan o pierden en las partidas. 

Conclusiones -> Sabiendo el rating de los jugadores de lichess, podemos ver en que rango de jugadores se pueden encontrar en la pagina,en este caso el jugador con más ranking al jugar con blancas es de 2700 puntos y el minimo de 784. Estos valores están lejos de la media de jugadores que se encuentra en 1596 puntos, la desviación estandar es de tan solo 291 puntos por lo que podemos concluir que la mayoria de los jugadores con piezas blancas están entre 1300-1800.

In [ ]:
pd.unique(data["black_rating"])

array([1191, 1261, 1500, ...,  840,  857,  860], dtype=int64)

In [ ]:
mediana_black_rating = data["black_rating"].median()
media_black_rating   = data["black_rating"].mean()
moda_black_rating    = data["black_rating"].mode().max()
std_black_rating     = data["black_rating"].std()
min_black_rating     = data["black_rating"].min()
max_black_rating     = data["black_rating"].max()

print("Mediana black_rating      :",mediana_black_rating)
print("Media black_rating        :",media_black_rating)
print("Moda black_rating         :",moda_black_rating)
print("Desv. Est. black_rating   :",std_black_rating)
print("Valor mínimo black_rating :",min_black_rating)
print("Valor máximo black_rating :",max_black_rating)

data["black_rating"].describe()

Mediana black_rating      : 1562.0
Media black_rating        : 1588.8319872370128
Moda black_rating         : 1500
Desv. Est. black_rating   : 291.03612596033696
Valor mínimo black_rating : 789
Valor máximo black_rating : 2723


count    20058.000000
mean      1588.831987
std        291.036126
min        789.000000
25%       1391.000000
50%       1562.000000
75%       1784.000000
max       2723.000000
Name: black_rating, dtype: float64

¿Qué representa? -> En esta sección se puede observar el puntaje de cada jugador que juegan con negras. Al igual que en el caso anterior, esto se ganan o pierden en las partidas. 

Conclusiones -> Aqui podemos ver que el jugador con más ranking al jugar con negras es de 2723 puntos y el minimo de 789. Estos valores están lejos de la media de jugadores que se encuentra en 1588 puntos, y en el mismo caso que las blancas, la desviación estandar es 291 puntos. Basandonos en el promedio, podemos decir que los jugadores de piezas negras tienen menor puntaje que los jugadores con piezas blancas.

In [ ]:
pd.unique(data["opening_name"])

array(['Slav Defense: Exchange Variation',
       'Nimzowitsch Defense: Kennedy Variation',
       "King's Pawn Game: Leonardis Variation", ...,
       "Queen's Indian Defense: Classical Variation |  Traditional Variation",
       "Queen's Gambit Declined: Slav |  Dutch Variation",
       'Sicilian Defense: Najdorf Variation |  Polugaevsky Variation'],
      dtype=object)

In [ ]:
# mediana_opening_name = data["opening_name"].median()
# media_opening_name  = data["opening_name"].mean()
moda_opening_name    = data["opening_name"].mode().max()
# std_opening_name     = data["opening_name"].std()
# min_opening_name     = data["opening_name"].min()
# max_opening_name     = data["opening_name"].max()

# print("Mediana opening_name      :",mediana_opening_name)
# print("Media opening_name        :",media_opening_name)
print("Moda opening_name         :",moda_opening_name)
# print("Desv. Est. opening_name   :",std_opening_name)
# print("Valor mínimo opening_name :",min_opening_name)
# print("Valor máximo opening_name :",max_opening_name)

data["opening_name"].describe() 

Moda opening_name         : Van't Kruijs Opening


count                    20058
unique                    1477
top       Van't Kruijs Opening
freq                       368
Name: opening_name, dtype: object

¿Qué representa? -> La variable "opening_name" nos dice la táctica de apertura que más veces usan al iniciar una partida de ajedrez.

Conclusiones -> Podemos decir que la táctica de apertura más usado en una partida de ajedrez es Van´t Kruijs.

In [ ]:
pd.unique(data["opening_ply"])

array([ 5,  4,  3, 10,  6,  1,  9,  2,  8,  7, 17, 11, 12, 13, 18, 19, 15,
       16, 14, 28, 20, 22, 24], dtype=int64)

In [ ]:
mediana_opening_ply = data["opening_ply"].median()
media_opening_ply  = data["opening_ply"].mean()
moda_opening_ply    = data["opening_ply"].mode().max()
std_opening_ply     = data["opening_ply"].std()
min_opening_ply     = data["opening_ply"].min()
max_opening_ply     = data["opening_ply"].max()

print("Mediana opening_ply      :",mediana_opening_ply)
print("Media opening_ply        :",media_opening_ply)
print("Moda opening_ply         :",moda_opening_ply)
print("Desv. Est. opening_ply   :",std_opening_ply)
print("Valor mínimo opening_ply :",min_opening_ply)
print("Valor máximo opening_ply :",max_opening_ply)

data["opening_ply"].describe() 

Mediana opening_ply      : 4.0
Media opening_ply        : 4.8169807558081565
Moda opening_ply         : 3
Desv. Est. opening_ply   : 2.797151810630167
Valor mínimo opening_ply : 1
Valor máximo opening_ply : 28


count    20058.000000
mean         4.816981
std          2.797152
min          1.000000
25%          3.000000
50%          4.000000
75%          6.000000
max         28.000000
Name: opening_ply, dtype: float64

¿Qué representa? -> La variable "opening_ply" nos dice el número de movimientos en la fase inicial del juego.

Conclusiones -> A partir de estos datos podemos decir que el promedio de movimientos en la fase inicial es de 4 movimientos. A su vez se tiene un valor mínimo de movimientos de 1 y un valor máximo de 28 movimientos en la fase inicial. Se tiene una desviación estandar de solo 2, por lo que podemos decir que se generan entre 2 y 6 movimientos en la fase inicial.